Userland/Applications/Dictionary/Dictionary.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2021, Robin Allen <r@foon.uk>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <AK/StringBuilder.h>
+
+#include "Dictionary.h"
+
+namespace Dictionary {
+
+static char const* part_of_speech_name(char pos)
+{
+    switch (pos) {
+    case 'n':
+        return "noun";
+    case 'v':
+        return "verb";
+    case 's':
+    case 'a':
+        return "adjective";
+    case 'r':
+        return "adverb";
+    default:
+        return "misc";
+    }
+}
+
+static int size_of_index_entry(char const* index_entry)
+{
+    int n = 0;
+    for (int i = 0; i < 16; i++) {
+        if (index_entry[i])
+            n++;
+        else
+            break;
+    }
+    return n;
+}
+
+Dictionary::Dictionary(char const* filename)
+{
+    auto file_or_error = MappedFile::map(filename);
+    if (file_or_error.is_error()) {
+        m_num_words = 0;
+        m_dictionary_bytes = 0;
+        return;
+    }
+
+    m_dictionary_file = file_or_error.value();
+    m_dictionary_bytes = (uint8_t*)m_dictionary_file->data();
+
+    Header const* header = (Header const*)m_dictionary_bytes;
+
+    m_num_words = header->num_words;
+    m_max_word_length = header->max_word_length;
+
+    m_index = offset_to_pointer(header->index_offset);
+    m_word_data_index = offset_to_pointer(header->word_data_index_offset);
+
+    outln("Number of words: {}", m_num_words);
+    outln("Max word length: {}", m_max_word_length);
+
+    VERIFY(m_num_words < 1000 * 1000);
+    VERIFY(m_max_word_length == 16);
+}
+
+StringView Dictionary::word_at_index(int word_index) const
+{
+    if (0 <= word_index && word_index < m_num_words) {
+        char const* c = (char const*)(m_index[word_index]);
+
+        int n = size_of_index_entry(c);
+
+        return StringView(c, n);
+    }
+    return "";
+}
+
+String Dictionary::definition_of(int word_index) const
+{
+    if (!(0 <= word_index && word_index < m_num_words)) {
```

Comment:
This construct seems kinda weird, might apply DeMorgan's?

```suggestion
    if (word_index < 0 || word_index >= m_num_words) {
```

Reply:
👍